### PR TITLE
Proposal: Convenience extension and subspec

### PR DESCRIPTION
**Pull request by mac-cain13**
**This PR contains:**
- Convenience subspec that depends on the iOS / R.swift subspecs
- `UIViewController` extension containing:
  - associated segue manager instance
  - swizzled method to provide default implementation for prepareForSegue calling them on the associated segue manager
  - proxy methods to perform segues

**The good:**
- Lazy initialization of the SegueManager
- `SegueManager` is kept alive by the `UIViewController` instance, so when the VC deallocs the manager goes with it
- Closer to what you're used to write
- Backwards compatible with vanilla iOS code, won't break anything
- Zero boilerplate, we go from this:

``` swift
import SegueManager

class MasterViewController: UIViewController {

  lazy var segueManager: SegueManager = {
    // SegueManager based on the current view controller
    return SegueManager(viewController: self)
  }()

  override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
    segueManager.prepareForSegue(segue)
  }

  func buttonTouched() {
    segueManager.performSegue(R.segue.someSegue) { _ in
      print("woohoo!")
    }
  }
}
```

to this:

``` swift
import SegueManager

class MasterViewController: UIViewController {
  func buttonTouched() {
    performSegue(R.segue.someSegue) { _ in
      print("woohoo!")
    }
  }
}
```

**The bad:**
- Uses ObjC associated values to save to retain the `SegueManager`, I don't consider this a problem, but pure Swift could be nicer. I tried using a caching system, but it is much more and error prone code with the same effect in the end...
- Uses method swizzling to implement the `prepareForSegue` method, this I think should be opt-in I believe. Apple docs state the default implementation does nothing, so we shouldn't break anything with this. But other libraries swizzling the same methods could give strange effects.
- Providing this as a subspec and having the R.swift compatible version as a subspec is a bit messy. Either the Convenience spec depends upon the R.swift spec or we have to create an extra ConvenienceWithR.swift spec...

**Conclusion:**
It feels really clean, nice and transparent to use as end user. We now often create all kinds of subclasses of ViewControllers in our projects. I think this could be a nice alternative. If it feels to ugly/risky to put into the main spec, then I would propose to provide this as a subspec.

_Todo:_
- [ ] Decide what to do with this proposal
- [ ] Update Readme/docs & example projects
- [ ] Create OS X version?!

### CODE REVIEW

File: SegueManager.podspec [hunk 1]
```diff
@@ -24,7 +24,14 @@ Pod::Spec.new do |s|
     ss.ios.deployment_target = '8.0'
     ss.source_files = "src/ios/SegueManager+Rswift.swift"
     ss.dependency "SegueManager/iOS"
-    ss.dependency "R.swift.Library"
+    ss.dependency "R.swift.Library", "~> 1.0.0.beta"
```
Comment (mac-cain13) [inline SegueManager.podspec:27]: Oh, I think I should revert this change. Only editted this so it would install the library beta. Should not be released. :)

**Comment by nmccann** [comment]
As a user of both R.swift and SegueManager I really like the convenience you're proposing, it has my vote :+1:

**Comment by tomlokhorst** [comment]
I've been thinking about this for a while, and I don't like the idea of encouraging these kinds of unsafe operations.

Slightly related: I was listening to [ATP's discussion](https://overcast.fm/+CdSzsTIY/38:14) on the swift-evolution proposal of making `final` the default for classes, and I like [Jordan Rose's point](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20151221/003897.html):

> Okay, so I probably shouldn't be putting this so bluntly, but the ship has already sailed on this. Supporting arbitrary code injection into someone else's framework is a non-goal for Swift, perhaps even an anti-goal.

I feel the same way about method swizzling. It provides weird behaviour (maybe other libraries are also swizzling the same method?). And it makes it hard to figure out how code works.
Whereas the current code may be verbose, but I don't think it's difficult to figure out. (See also: [Clarity is more important than brevity](https://swift.org/documentation/api-design-guidelines.html#clarity-over-brevity))

---

Having said all that, I do like the idea of making SegueManager more convenient to use. I've been thinking of several ways:

**1. Provide a super class with default implementation**

The SegueManger library can ship a default super class:

``` swift
public class ViewController : UIViewController {
  public lazy var segueManager: SegueManager = { return SegueManager(viewController: self) }()

  public override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
    segueManager.prepareForSegue(segue)
  }
}
```

Now all you have to do is subclass this, and you're done!

I personally would not use this, since I don't like giving up my only inheritance path to libraries. But I guess others don't have this problem.

**2. Provide a protocol**

Can't really think of a good name right now, but something like:

``` swift
public protocol Segueable {
  var segueManager: SegueManager { get }
}
```

Of course then we can provide a protocol extension that uses that:

``` swift
public extension Segueable {
  func performSegue<Segue, Source, Destination>(
    segueIdentifier: StoryboardSegueIdentifier<Segue, Source, Destination>,
    handler: TypedStoryboardSegueInfo<Segue, Source, Destination> -> Void)
  {
    segueManager.performSegue(segueIdentifier, handler: handler)
  }
}
```

And make the ViewController super class confirm to that protocol:

``` swift
extension ViewController : Segueable {}
```

Users can also make their own classes confirm to this protocol.
This way you also get access to the convenience method:

``` swift
  @IBAction func openDetailAction(sender: UIButton) {

    self.performSegue(R.segue.masterViewController.showDetail) { segue in
      let detailController = segue.destinationViewController
      detailController.displayText = "This is the detail screen!"
    }
  }
```

**Comment by mac-cain13** [comment]
Yeah, I agree method swizzling and even using associated object is not really a thing that is nice to do from a lib. It can be very confusing especially when libraries conflict etc.

I think combining 1 and 2 from your proposal is nice. You can use whatever you like. It doesn't force you to use anything, but you can make use of it quite easily. Up to the developer to decide whether to give up their inheritance option to the lib.

That said, it really feels like something Apple should fix in UIKit, the swizzle solution really feels like what Apple should have done in the first place.

**Comment by tomlokhorst** [comment]
The current develop branch has a `SegueManagerViewController`,  `SegueManagerTableViewController` and related with a default implementation. As well as a `HasSegueManager` protocol including a `performSegue` extension for R.swift.

This should cover most of the points discussed.
